hpen.c

```diff
@@ -0,0 +1,491 @@
+	}
+
+	/* Parse inputs */
+	hd = hid_start_parse(d_ptr, d_len, 1 << hid_input);
+	HID_TLC_FOREACH_ITEM(hd, &hi, tlc_index) {
+		switch (hi.kind) {
+		case hid_collection:
+			if (hi.collevel == 1 && hi.usage ==
+			    HID_USAGE2(HUP_DIGITIZERS, HUD_PEN)) {
+				pen_coll = true;
+				pen_found = true;
+			}
+			break;
+		case hid_endcollection:
+			if (hi.collevel == 0 && pen_coll)
+				pen_coll = false;
+			break;
+		case hid_input:
+			if (!sc || !pen_coll)
+				break;
+			sc->report_id = hi.report_ID;
+			for (size_t i = 0; i < HPEN_N_USAGES_ABS; i++) {
+				if (hi.usage == hpen_hid_map_abs[i].usage) {
+					bit_set(sc->abs_caps, i);
+					sc->locs_abs[i] = hi.loc;
+					sc->ai[i] = (struct hpen_absinfo) {
+					    .max = hi.logical_maximum,
+					    .min = hi.logical_minimum,
+					    .res = hid_item_resolution(&hi),
+					};
+				}
+			}
+			for (size_t i = 0; i < HPEN_N_USAGES_KEY; i++) {
+				if (hi.usage == hpen_hid_map_key[i].usage) {
+					bit_set(sc->key_caps, i);
+					sc->locs_key[i] = hi.loc;
+				}
+			}
+			break;
+		default:
+			break;
+		}
+	}
+	hid_end_parse(hd);
+
+	if (!pen_found)
+		return (ENXIO);
+
+	if (!sc)
+		return (0);
+
+	for (i = 0; i < HPEN_N_USAGES_ABS; i++) {
+		if (hpen_hid_map_abs[i].required && !USAGE_SUPPORTED(sc->abs_caps, i)) {
+			device_printf(sc->sc_dev, "required report %s not found\n", hpen_hid_map_abs[i].name);
+			return (ENXIO);
+		}
+	}
+
+	for (i = 0; i < HPEN_N_USAGES_KEY; i++) {
+		if (hpen_hid_map_key[i].required && !USAGE_SUPPORTED(sc->key_caps, i)) {
+			device_printf(sc->sc_dev, "required report %s not found\n", hpen_hid_map_abs[i].name);
+			return (ENXIO);
+		}
+	}
+
+	sc->isize = hid_report_size_1(d_ptr, d_len, hid_input, sc->report_id);
+
+	return (0);
+}
+
+static int
+hpen_probe(device_t dev)
+{
+	struct hpen_softc *sc = device_get_softc(dev);
+	void *d_ptr;
+	uint16_t d_len;
+	int error;
+
+	if (hidbus_get_usage(dev) != HID_USAGE2(HUP_DIGITIZERS, HUD_PEN))
+		return (ENXIO);
+
+	error = hid_get_report_descr(dev, &d_ptr, &d_len);
+	if (error != 0) {
+		device_printf(dev, "could not retrieve report descriptor from "
+		     "device: %d\n", error);
+		return (ENXIO);
+	}
+
+	/* Check if report descriptor belongs to a HID tablet device */
+	if (hpen_hid_parse(sc, d_ptr, d_len,
+		    hidbus_get_usage(dev), hidbus_get_index(dev)) != 0)
+		return (ENXIO);
+
+	return (BUS_PROBE_DEFAULT);
+}
+
+static int
+hpen_attach(device_t dev)
+{
+	struct hpen_softc *sc = device_get_softc(dev);
+	struct hid_device_info *hw = device_get_ivars(dev);
+	void *d_ptr = NULL;
+	int err;
+	uint16_t d_len;
+	size_t i;
+
+	sc->sc_dev = dev;
+	device_set_desc(dev, hw->name);
+
+	/* Get HID descriptor */
+	err = hid_get_report_descr(dev, &d_ptr, &d_len);
+	if (err) {
+		device_printf(dev, "could not retrieve report descriptor from "
+				 "device: %d\n", err);
+		return (ENXIO);
+	}
+
+	hpen_hid_parse(sc, d_ptr, d_len, hidbus_get_usage(dev), hidbus_get_index(dev));
```

Comment:
report descriptor has been parsed in probe method already

Reply:
Right, I thought I couldn't use the `sc` in `probe`, hence the double parsing, without `sc` in `probe` and with `sc` in `attach`. But looks like that does work.